### ISSUE TITLE
Abort on startup: `CommandLine Error: Option ... registered more than once!`

### ISSUE THREAD

**Issue by fbs** [bpftrace/bpftrace#1855]
Bcc static links llvm by default, if we then link LLVM again we get `llvm command line` fatal errors during startup.

We should either:

- find out whether we can fix this at link time. Maybe if we specify bcc before llvm it grabs most of the stuff it needs from the bcc archive, preventing issues. Although it seems tricky to get right
- Note in the docs that bcc needs to be built with shared llvm, see:

---
Did you build bcc yourself, if you did make sure you dynamic link against llvm. See debc79ef9ad4784258705a92ae70f9c7689a9c24

_Originally posted by @fbs in https://github.com/iovisor/bpftrace/issues/251#issuecomment-846424774_

---

: CommandLine Error: Option 'openmp-ir-builder-optimistic-attributes' registered more than once!

**Comment by Rtoax** [comment]
> This is only known to be an issue with LLVM 12.
> 
> The current suggested workaround is to build bcc with: `-DENABLE_LLVM_SHARED=1` cmake flag

`-DENABLE_LLVM_SHARED=1` solve my problem.

**Comment by brendangregg** [comment]
The bcc with `-DENABLE_LLVM_SHARED=1` fix worked for me, thanks!

**Comment by Rtoax** [comment]
> 

Me too.

**Comment by Rtoax** [comment]
The bcc with `-DENABLE_LLVM_SHARED=1` fix worked for me, thanks!

**Comment by Jongy** [comment]
The bcc with `-DENABLE_LLVM_SHARED=1` fix worked for me, thanks!

**Comment by ILXR** [comment]
Thanks very much! It saved me.

**Comment by Rtoax** [comment]
> Thanks very much! It saved me.

Are you `dnf up` on centos stream9?

**Comment by ILXR** [comment]
> 

No, I use apt on Ubuntu 20

**Comment by ILXR** [comment]
> > Thanks very much! It saved me.
> 
> Are you `dnf up` on centos stream9?

No, I use apt on Ubuntu 20

**Comment by liutingjieni** [comment]
I also encountered this problem when compiling bpftrace after reinstalling bcc. My solution is to delete the build directory of bpftrace and recompile (cmake, make, make install)

**Comment by haomi123** [comment]
The bcc with -DENABLE_LLVM_SHARED=1  doesn't solve my problem

: CommandLine Error: Option 'filter' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

**Comment by NirLevy98** [comment]
The bcc with -DENABLE_LLVM_SHARED=1 doesn't solve my problem

: CommandLine Error: Option 'help-list' registered more than once!

**Comment by matthew-olson-intel** [comment]
Using Arch with LLVM version:
```
$ llvm-config --version
14.0.6
```

Recompiling `bcc` with `-DENABLE_LLVM_SHARED=1` fixed it for me, as well.

**Comment by marcosdotps** [comment]
Ubuntu 22.04:

```bash
$ llvm-config --version
14.0.0
```
Recompiling with `-DENABLE_LLVM_SHARED=1` has not fixed the problem for me 🤔

**Comment by danobi** [comment]
To be clear, `-DENABLE_LLVM_SHARED=1` is intended for bcc, not bpftrace.

**Comment by gukoff** [comment]
For me, the error was caused by the bcc installed from source. I fixed it with:

```
cd bcc/build
sudo make uninstall

sudo apt install bpftrace bpfcc-tools libbpfcc  --reinstall
```

**Comment by dnb02** [comment]
> This is only known to be an issue with LLVM 12.
> 
> The current suggested workaround is to build bcc with: `-DENABLE_LLVM_SHARED=1` cmake flag


@danobi just to make it clear, when building bcc we will do 
```
git clone https://github.com/iovisor/bcc.git
mkdir bcc/build; cd bcc/build
cmake ..
make
sudo make install
cmake -DPYTHON_CMD=python3 -DENABLE_LLVM_SHARED=1..
pushd src/python/
make
sudo make install
popd
```

**Comment by devoli170** [comment]
I get a similar error on ubuntu 24.10 (oracular). I installed the bpftrace package and all dependencies using apt.

```
apt info bpftrace
Package: bpftrace
Version: 0.21.2-1ubuntu1
```

```
bpftrace --version
: CommandLine Error: Option 'debug-counter' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
Aborted (core dumped)
```

running bpftrace in the focal container seems fine

```
docker run -it ubuntu:focal /bin/sh -c "apt update; apt install -y bpftrace; bpftrace --version"
...
bpftrace v0.9.4
```

with the oracular container i can reproduce the issue i have on my host:
```
docker run -it ubuntu:oracular /bin/sh -c "apt update; apt install -y bpftrace; bpftrace"
...
: CommandLine Error: Option 'debug-counter' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
Aborted (core dumped)
```

I'm confused about the bcc dependency situation.  According to https://github.com/bpftrace/bpftrace/pull/2265 it should be removed, right? 

it's replaced by libbcc in focal:

```
docker run -it ubuntu:focal /bin/sh -c "apt update; apt install -y bpftrace;ldd $(which bpftrace) | grep bcc"
...
	libbcc.so.0 => /lib/x86_64-linux-gnu/libbcc.so.0 (0x000078e1d4400000)
```

however it's back in oracular:
```
docker run -it ubuntu:oracular /bin/sh -c "apt update; apt install -y bpftrace; ldd $(which bpftrace) | grep bcc"
...
		libbcc_bpf.so.0 => /lib/x86_64-linux-gnu/libbcc_bpf.so.0 (0x00007cf23ad98000)
```

**Comment by jordalgo** [comment]
> I'm confused about the bcc dependency situation. According to https://github.com/bpftrace/bpftrace/pull/2265 it should be removed, right?

Unfortunately, we still use bcc for some things e.g. symbolication and some probe lookup/attaching but we're hoping to get rid of it completely by end of next year.

@devoli170 As far as the crash goes, that looks like an LLVM issue. What version of LLVM are you on? Maybe try upgrading it past 12.

**Comment by devoli170** [comment]
```
llvm-cov --version
Ubuntu LLVM version 19.1.1
  Optimized build.
```

however the bpftrace package depends on libllvm18, so this lib is installed alongside the llvm-19 packages. bpftrace also uses `libLLVM-18.so.18.1 => /lib/x86_64-linux-gnu/libLLVM-18.so.18.1`

**Comment by roopeshsn** [comment]
Hello everyone! My Ubuntu release is 22.04. I installed `bcc` from the source (without the flag `-DENABLE_LLVM_SHARED=1`) and then `bpftrace` from `apt`. What I've noticed through the command, `dpkg -l | grep llvm` is that they both installing different releases of LLVM (11, and 13). Should I need to uninstall `bcc` and build it again using the `DENABLE_LLVM_SHARED=1` flag and also installing `bpftrace` from the source?

**Comment by kukrimate** [comment]
On Ubuntu this is caused by having two copies of LLVM in the same process.
- bpfcc is linked to LLVM dynamically as expected.
- bpftrace pulls in the _static_ libraries, despite Ubuntu's build process not asking for it.

This causes to copies of LLVM to be loaded into the resulting bpftrace process.

After some investigation, this is due to the following check failing in `src/ast/CMakeLists.txt`:
```
  find_library(found_LLVM LLVM HINTS ${LLVM_LIBRARY_DIRS})
  if(found_LLVM)
```

This is despite find_package correctly finding our LLVM. Removing this check and always dynamically linking LLVM if STATIC_LINKING=OFF` seems to fix the issue.

To users in this thread, the fix will be available in Ubuntu 24.10 as soon as feasible.
